### PR TITLE
DRA: fix data race

### LINKED ISSUE #123474: DRA: kubelet dies with "concurrent map iteration and map write"
### What happened?

I was running `_output/bin/ginkgo -p -focus="DynamicResourceAllocation" ./test/e2e` with a local-up-cluster.sh cluster when kubelet died.

Here's the output:
```
I0223 18:04:20.818486  387925 round_trippers.go:553] GET https://localhost:6444/apis/resource.k8s.io/v1alpha2/namespaces/dra-2883/resourceclaims/external-claim 200 OK in 4 milliseconds
I0223 18:04:20.819006  387925 round_trippers.go:553] GET https://localhost:6444/apis/resource.k8s.io/v1alpha2/namespaces/dra-2883/resourceclaims/external-claim 200 OK in 6 milliseconds
I0223 18:04:20.819223  387925 round_trippers.go:553] GET https://localhost:6444/apis/resource.k8s.io/v1alpha2/namespaces/dra-2883/resourceclaims/external-claim 200 OK in 7 milliseconds
I0223 18:04:20.819258  387925 round_trippers.go:553] GET https://localhost:6444/apis/resource.k8s.io/v1alpha2/namespaces/dra-2922/resourceclaims/tester-1-my-inline-claim-m4q4f 200 OK in 7 milliseconds
fatal error: concurrent map iteration and map write

goroutine 51298 [running]:
reflect.mapiternext(0x4f85c9?)
        runtime/map.go:1392 +0x13
reflect.(*MapIter).Next(0xc008e711c0?)
        reflect/value.go:2005 +0x74
encoding/json.mapEncoder.encode({0x7f6120609460?}, 0xc00c6a0840, {0x3d33520?, 0xc001f11f60?, 0x33bb960?}, {0xa?, 0x0?})
        encoding/json/encode.go:745 +0x334
encoding/json.structEncoder.encode({{{0xc000998908, 0x8, 0x8}, 0xc000db00f0, 0xc000db0120}}, 0xc00c6a0840, {0x3c4d040?, 0xc001f11f10?, 0x0?}, {0x0, ...})
        encoding/json/encode.go:704 +0x21e
encoding/json.arrayEncoder.encode({0x7f6172cb4108?}, 0xc00c6a0840, {0x34474e0?, 0xc0067ff150?, 0x0?}, {0x0?, 0x0?})
        encoding/json/encode.go:847 +0xcf
encoding/json.sliceEncoder.encode({0x50ac65?}, 0xc00c6a0840, {0x34474e0?, 0xc0067ff150?, 0x33bb960?}, {0xa?, 0x0?})
        encoding/json/encode.go:820 +0x347
encoding/json.structEncoder.encode({{{0xc00066f448, 0x3, 0x4}, 0xc000db01b0, 0xc000db01e0}}, 0xc00c6a0840, {0x393ec60?, 0xc0067ff140?, 0xc00a0091d0?}, {0x0, ...})
        encoding/json/encode.go:704 +0x21e
encoding/json.(*encodeState).reflectValue(0xc00c6a0840, {0x393ec60?, 0xc0067ff140?, 0xc008e71830?}, {0xe0?, 0x9d?})
        encoding/json/encode.go:321 +0x73
encoding/json.(*encodeState).marshal(0x41535b?, {0x393ec60?, 0xc0067ff140?}, {0xd0?, 0x91?})
        encoding/json/encode.go:297 +0xc5
encoding/json.Marshal({0x393ec60, 0xc0067ff140})
        encoding/json/encode.go:163 +0xd0
k8s.io/kubernetes/pkg/kubelet/cm/dra/state.(*DRAManagerCheckpoint).MarshalCheckpoint(0xc0067ff050)
        k8s.io/kubernetes/pkg/kubelet/cm/dra/state/checkpoint.go:69 +0x4e
k8s.io/kubernetes/pkg/kubelet/checkpointmanager.(*impl).CreateCheckpoint(0xc000ab3ad0, {0x3e85742, 0x11}, {0x4433e40?, 0xc0067ff050?})
        k8s.io/kubernetes/pkg/kubelet/checkpointmanager/checkpoint_manager.go:69 +0xc9
k8s.io/kubernetes/pkg/kubelet/cm/dra/state.(*stateCheckpoint).store(0xc000a17180, {0xc001f11808, 0x11, 0x11})
        k8s.io/kubernetes/pkg/kubelet/cm/dra/state/state_checkpoint.go:147 +0xb7
k8s.io/kubernetes/pkg/kubelet/cm/dra/state.(*stateCheckpoint).Store(0xc000a17180, {0xc001f11808, 0x11, 0x11})
        k8s.io/kubernetes/pkg/kubelet/cm/dra/state/state_checkpoint.go:139 +0x7e
k8s.io/kubernetes/pkg/kubelet/cm/dra.(*claimInfoCache).syncToCheckpoint(0xc000db02d0)
        k8s.io/kubernetes/pkg/kubelet/cm/dra/claiminfo.go:222 +0x268
k8s.io/kubernetes/pkg/kubelet/cm/dra.(*ManagerImpl).PrepareResources(0xc000c030c8, 0xc004e0e008)
        k8s.io/kubernetes/pkg/kubelet/cm/dra/manager.go:214 +0xbe5
k8s.io/kubernetes/pkg/kubelet/cm.(*containerManagerImpl).PrepareDynamicResources(0x0?, 0x1?)
        k8s.io/kubernetes/pkg/kubelet/cm/container_manager_linux.go:1017 +0x22
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).PrepareDynamicResources(0xc0006283c0?, 0x3e9bf34?)
        k8s.io/kubernetes/pkg/kubelet/kubelet.go:3039 +0x25
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod(0xc000a46300, {0x4436f20, 0xc00c01edf0}, 0xc004e0e008, 0xc00b914090, {0x62e0fa0, 0x0, 0x0}, 0xc000287db0)
        k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:1145 +0x1762
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).SyncPod(0xc00096d808, {0x4436ee8, 0xc0079a19f0}, 0x2, 0xc004e0e008, 0x0, 0xc00b914090)
        k8s.io/kubernetes/pkg/kubelet/kubelet.go:1955 +0x2a83
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop.func1({0x0, {0x2, {0xc16e51e5189c05b9, 0x1b1d167c2fb, 0x627e340}, 0xc004e0e008, 0x0, 0x0, 0x0}}, 0xc000a01900, ...)
        k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1283 +0x1ca
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop(0xc000a01900, {0xc002d5b6e0, 0x24}, 0xc002518de0)
        k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1288 +0x49b
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1()
        k8s.io/kubernetes/pkg/kubelet/pod_workers.go:950 +0x118
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod in goroutine 351
        k8s.io/kubernetes/pkg/kubelet/pod_workers.go:945 +0x20db

...
```

/assign @bart0sh 
/sig node
cc @klueska 

### What did you expect to happen?

kubelet should have kept running.

### How can we reproduce it (as minimally and precisely as possible)?

Run DRA E2E tests. It might be necessary to use my `dra-structured-parameters` branch, it has some additional E2E tests, some of which are stressing kubelet more than before.

### Anything else we need to know?

_No response_

### Kubernetes version

Master + local changes.

### Cloud provider

n.a.

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by bart0sh**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR fixes data race caused by these 2 actions performed at the same time:
- `PrepareResources` reads cached `ClaimInfos` when checkpointing
- `UnprepareResources` modifies them when deleting pod reference

There could be other scenarios triggering this type of data races.
    
The fix is to Lock `ClaimInfo` and deep copy it when checkpointing.

#### Which issue(s) this PR fixes:

Fixes #123474

#### Special notes for your reviewer:

New DRA Manager test case `TestParallelPrepareUnprepareResources` helped to reproduce the issue.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/cm/dra/claiminfo.go [hunk 1]
```diff
@@ -216,7 +216,10 @@ func (cache *claimInfoCache) syncToCheckpoint() error {
 
 	claimInfoStateList := make(state.ClaimInfoStateList, 0, len(cache.claimInfo))
 	for _, infoClaim := range cache.claimInfo {
-		claimInfoStateList = append(claimInfoStateList, infoClaim.ClaimInfoState)
+		infoClaim.Lock()
```
Comment (pohly) [inline pkg/kubelet/cm/dra/claiminfo.go:219]: How is access to `cache.claimInfo` protected against concurrent access?

Comment (bart0sh) [reply pkg/kubelet/cm/dra/claiminfo.go:219]: by the cache.RLock() on line 214 above

Comment (pohly) [reply pkg/kubelet/cm/dra/claiminfo.go:219]: Do we really need the `ClaimInfo.RWMutex`? In other words, are there code paths which get a `ClaimInfo` pointer and then modify that instance under a write lock while not holding the `claimInfoCache? My concern is that taking a lock on the cache is not a guarantee to get a consistent snapshot of the cache content.

I found https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/pkg/kubelet/cm/container_manager_linux.go#L955-L982 but that only reads. `GetContainerClaimInfos` could be changed to return a deep-copy of the information, then the caller can use that without locking.

If we need this two-level locking, can you add source code comments explaining the expected usage of these structs and their mutexes?

Comment (pohly) [reply pkg/kubelet/cm/dra/claiminfo.go:219]: For example, take this code:

https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/pkg/kubelet/cm/dra/manager.go#L111-L126

What could happen is:
- One PrepareResources call starts, gets to the point where it has added a pod reference to the claim info but not actually invoked NodePrepareResources.
- Another PrepareResources starts and completes, thus flushing the cache via `m.cache.syncToCheckpoint()`.

=> An inconsistent cache gets written, right?

Comment (bart0sh) [reply pkg/kubelet/cm/dra/claiminfo.go:219]: @pohly [here](https://github.com/kubernetes/kubernetes/pull/124323) is an attempt to remove locking from the `ClaimInfo` and ensure that cache objects references are not used directly, PTAL.

**Comment by bart0sh** [comment]
/close
in favour of https://github.com/kubernetes/kubernetes/pull/124323

**Pull request closed without merging**